src/differentials/composite.jl

```diff
@@ -82,7 +82,16 @@ end
 Base.keys(comp::Composite) = keys(backing(comp))
 Base.propertynames(comp::Composite) = propertynames(backing(comp))
 
-Base.iterate(comp::Composite, args...) = iterate(backing(comp), args...)
+Base.iterate(comp::Composite) = iterate(comp, 1)
+
+function Base.iterate(comp::Composite, state::Integer)
+    if state <= length(comp)
+        return (getindex(comp, state), state + 1)
+    else
+        return nothing
+    end
+end
+
```

Comment:
Or I can replace the entire code block with:
```jl
function Base.iterate(comp::Composite, args...)
    out = iterate(backing(comp), args...)
    if out isa Nothing
        return out
    else
        return (unthunk(out[1]), out[2])
    end
end
```

Reply:
If I look at the benchmarks, the current code in the pull request gives me:

```jl
julia> @btime iterate(r, 2)
  47.447 ns (1 allocation: 32 bytes)
```
while the suggested code gives me:
```jl
julia> @btime iterate(r, 2)
  224.165 ns (2 allocations: 48 bytes)
```

Reply:
my instinct is the suggested code is better. I will take a look into why it is slower

Reply:
Poking at this, it looks like something funky is happening where it is deciding not to constant fold the state in the iterate.
I get on julia-master no allocations and instant on both.

but when timing: `[x for x in $r]` the getindex code has 1 allocation and the iterate backing code has 2 allocations.

Weird.
But i guess this is fine

Reply:
Ok, I've used the `iterate` approach then.

Reply:
I mean the getindex appraoch is fine